### ISSUE TITLE
getting occasional EPIPE error

### ISSUE THREAD

**Issue by bhumilsarvaiya** [neo4j/neo4j-javascript-driver#375]
Hello,
I am currently using Neo4j v3.3.4 with neo4j javascript driver v1.5.3 and node v9.8.0. Neo4j database is running inside a docker container and my node app in another docker container. They are deployed using swarm mode on google instance. I am getting this error at regular interval.

{"stack":"Neo4jError: write EPIPE\n\n    at captureStacktrace (/home/app/node_modules/neo4j-driver/lib/v1/result.js:200:15)\n    at new Result (/home/app/node_modules/neo4j-driver/lib/v1/result.js:73:19)\n    at _runPullAll (/home/app/node_modules/neo4j-driver/lib/v1/transaction.js:338:10)\n    at Object.commit (/home/app/node_modules/neo4j-driver/lib/v1/transaction.js:250:24)\n    at Transaction.commit (/home/app/node_modules/neo4j-driver/lib/v1/transaction.js:123:35)\n    at /home/app/node_modules/neo4j-driver/lib/v1/internal/transaction-executor.js:126:16\n    at <anonymous>\n    at process._tickCallback (internal/process/next_tick.js:118:7)","level":"info","message":"write EPIPE","timestamp":"2018-05-23T07:53:19.890Z"}

My TypeScript code for connection and executing query is
```js
class Neo4jClient {
    static instance: Neo4jClient | undefined;
    client: neo4j.Driver | undefined;
    constructor () {
        try {
            this.client = neo4j.driver("bolt://neo4j", neo4j.auth.basic("neo4j", password));
        } catch (err) {
            logging.logServerError(err);
        }
    }
}

function getClient(): neo4j.Driver | undefined {
    if (Neo4jClient.instance === undefined) {
        Neo4jClient.instance = new Neo4jClient();
    }
    return Neo4jClient.instance.client;
}

type paramTypes = { [key: string]: any };

export function executeRead(query: string, params: paramTypes): Promise<neo4j.StatementResult> {
    return new Promise((resolve, reject) => {
        let client = getClient();
        if (client === undefined) {
            Neo4jClient.instance = undefined;
            throw Error("neo4j not connected");
        }
        let session = client.session();
        session.readTransaction((transaction: neo4j.Transaction) => {
            transaction.run(query, params)
            .then((result: neo4j.StatementResult) => {
                session.close();
                resolve(result);
            })
            .catch ((err: Error) => {
                reject(err);
            });
        });
    });
}

export function executeWrite(query: string, params: paramTypes): Promise<neo4j.StatementResult> {
    return new Promise((resolve, reject) => {
        let client = getClient();
        if (client === undefined) {
            Neo4jClient.instance = undefined;
            throw Error("neo4j not connected");
        }
        let session = client.session();
        session.writeTransaction((transaction: neo4j.Transaction) => {
            transaction.run(query, params)
            .then((result: neo4j.StatementResult) => {
                session.close();
                resolve(result);
            })
            .catch ((err: Error) => {
                reject(err);
            });
        });
    });
}
```

I am calling **executeWrite** and **executeRead** in my APIs.

**Comment by lutovich** [comment]
Hello @bhumilsarvaiya,

I think EPIPE error happens when connection used by the driver has been closed by the other side. It might be closed either by the database or by the network infrastructure. The database should always log such close events in `neo4j.log` or `debug.log`. Seeing them would be useful. Network infra might close idle connections after a timeout. Do you know if such a timeout exists in your setup?

Right now I can recommend tunning `maxConnectionLifetime` setting to make driver recycle socket connections. It should be set to a lower value than the existing idle connection timeout in the networking infra. More details about this can be found in the [manual](https://neo4j.com/docs/developer-manual/current/drivers/client-applications/#driver-config-connection-pool-management).

It is also worth changing the place where sessions are closed. They should be closed after the transaction function completes to not interfere with possible retries:

```typescript
export function executeRead(query: string, params: paramTypes): Promise<neo4j.StatementResult> {
  return new Promise((resolve, reject) => {
    let client = getClient();
    if (client === undefined) {
      Neo4jClient.instance = undefined;
      throw Error("neo4j not connected");
    }
    let session = client.session();
    session.readTransaction((transaction: neo4j.Transaction) => transaction.run(query, params))
      .then((result: neo4j.StatementResult) => {
        session.close();
        resolve(result);
      })
      .catch((err: Error) => {
        reject(err);
      });
  });
}
```

Hope this helps!

**Comment by lutovich** [comment]
Hi @bhumilsarvaiya,

Any updates on this issue?

**Comment by bhumilsarvaiya** [comment]
Hi @lutovich,

Thank you very much for the help. I went through few of your other issues and found something regarding Google Instance closing idle TCP connection after 10 mins. Ideally that should not be the reason for my problem as I am using docker swarm overlay network which should not have closed the connection (I am using few other databases and not getting any TCP related problems). But still as suggested in that issue as well as in this one, I changed ***maxConnectionLifetime*** to ***10 mins*** and since then I am not getting the EPIPE error. Also I have changed the way I close the session as per your recommendation. But I don't think that solved the problem as I made that change few days later after making ***maxConnectionLifetime*** change and by then I already stopped getting EPIPE errors.

Closing the issue now. Will reopen if I ever get the error again.

Thank you :v:
